### PR TITLE
HMT 35: Link users to teams

### PR DESCRIPTION
**Pull request by AnaCoda**
Created a lambda function that takes in two arguments: **teamId**, **userId** and then adds the corresponding user to the team.

The following return values are possible:

- 404: "Error: Team does not exist"
- 404: "Error: User does not exist"
- 400: "Error: User is already part of a team"
- 400: "Error: Team is full"
- 200: "Success" (User has been connected to the team)

### CODE REVIEW

File: amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { defineFunction } from "@aws-amplify/backend";
+
+export const AssignUsersToTeams = defineFunction({
+  entry: "./handler.ts",
+  runtime: 20,
+  name: "AssignUsersToTeams",
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts:6]: I know this was originally empty for defaults, I just filled these out since it helps for code clarity when troubleshooting.

File: amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  let team;
+  try {
+    team = await dataClient.models.Team.get({ id: event.arguments.teamId });
+  } catch (err) {
+    return {
+      body: { value: "Error: Team does not exist" },
+      statusCode: 404,
+      headers: { "Content-Type": "application/json" },
+    };
+  }
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts:64]: let's make this check be like the user one for two reasons:
1. This won't throw/catch if ID isn't found since GraphQL returns a successful promise with an errors:[] field.
2. Just to be consistent

File: amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  if (await user.data.Team()) {
+    return {
+      body: { value: "Error: User is already part of a team" },
+      statusCode: 400,
+      headers: { "Content-Type": "application/json" },
+    };
+  }
+
+  await dataClient.models.User.update({
+    id: event.arguments.userId,
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts:85]: Assign this to a variable and check the result does not contain any errors ie `(!result.errors)` before returning status code 200. We wan't to ensure 200 is not returned if an error occurs here.

File: amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    TeamId: team.data.id,
+  });
+
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts:88]: Let's extract these all to a const variable before the if for consistency and reusability.

**Review by ideen1 — CHANGES_REQUESTED** [review]
Looks great over all awesome work! Just a few fixes. Once these are pushed we can do some final testing and merge.

File: amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts (hunk 1 above)
Comment (ideen1) [reply amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts]: This comment still needs to have resource information populated. It is also important that we set Node Version to 20 to ensure we are using a runtime that will be supported for a few years at least.

File: amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { defineFunction } from "@aws-amplify/backend";
+
+export const AssignUsersToTeams = defineFunction({
+  runtime: 20,
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/resource.ts:4]: Let's add the entry and name parameters here. It's true that entry is defaulted to handler, but someone who is new to Amplify or is not familiar with this module will have more difficulty understanding what's going on.

File: amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    if (user == null) {
+      return {
+        body: { value: "Error: User does not exist" },
+        statusCode: 404,
+        headers: { "Content-Type": "application/json" },
+      };
+    }
+
+    if (await user.team) {
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts:90]: Because of the way we made the graphQL query, these all return as part of the main await so we no longer need to await any of these. We can remove this await

File: amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      return {
+        body: { value: "Error: User is already part of a team" },
+        statusCode: 400,
+        headers: { "Content-Type": "application/json" },
+      };
+    }
+
+    const membersConnection = (await team.members) as ModelUserConnection;
```
Comment (ideen1) [inline amplify/function/BusinessLogic/AssignUsersToTeams/handler.ts:98]: Same as the above we can remove this and the one in the line below and actually replace it with something like:

```suggestion
    const members = team.members.items as ModelUserConnection;
```

File: amplify/data/resource.ts [hunk 8]
```diff
@@ -59,9 +60,20 @@ const schema = a
       // allow all users to call this api for now
       .authorization((allow) => [allow.guest()])
       .handler(a.handler.function(DemoFunction)),
+    AssignUsersToTeams: a
+      .mutation()
+      .arguments({
+        userId: a.string(),
```
Comment (ideen1) [inline amplify/data/resource.ts:66]: Let's make these two arguments required with `.required()`

**Review by ideen1 — CHANGES_REQUESTED** [review]
Looks great, also good work dealing with all the Amplify migration changes. Just a few small changes and then we can merge it in. I also re-tested all edge cases and it seems to response as expected!

**Review by ideen1 — APPROVED** [review]
Looks good
